File: server/ruby/tests/test_all.rb
#!/usr/bin/env ruby
# Test script to validate all Ruby code samples.
# This script checks:
# 1. Ruby syntax validity
# 2. Import dependencies
# 3. Basic structure validation

require 'fileutils'

# Colors for output
GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
RESET = "\033[0m"

def check_syntax(file_path)
  begin
    # Use Ruby's syntax checker
    result = `ruby -c "#{file_path}" 2>&1`
    if $?.success?
      { ok: true, message: "✓ Syntax valid" }
    else
      { ok: false, message: "✗ Syntax error: #{result.chomp}" }
    end
  rescue => e
    { ok: false, message: "✗ Error checking syntax: #{e.message}" }
  end
end

def check_imports(file_path)
  begin
    content = File.read(file_path)
    
    # Check for required imports
    has_dotenv = content.include?('dotenv') || content.include?("require 'dotenv'")
    has_json = content.include?('json') || content.include?("require 'json'")
    has_net_http = content.include?('net/http') || content.include?("require 'net/http'")
    
    missing = []
    missing << 'dotenv' unless has_dotenv
    missing << 'json' unless has_json
    missing << 'net/http' unless has_net_http
    
    if missing.any?
      { ok: false, message: "✗ Missing requires: #{missing.join(', ')}" }
    else
      # Try to check if gems are available
      begin
        require 'dotenv'
        require 'json'
        require 'net/http'
        { ok: true, message: "✓ Imports valid" }
      rescue LoadError => e
        { ok: false, message: "✗ Dependencies not installed (run: bundle install)" }
      end
    end
  rescue => e
    { ok: false, message: "✗ Error checking imports: #{e.message}" }
  end
end

def check_structure(file_path)
  begin
    content = File.read(file_path)
    
    checks = []
    checks << "Environment loading" if content.include?('Dotenv.load') || content.include?('dotenv')
    checks << "API key reference" if content.include?('COURIER_API_KEY')
    checks << "HTTP client" if content.include?('Net::HTTP') || content.include?('net/http')
    checks << "JSON handling" if content.include?('JSON') || content.include?('json')
    checks << "Authorization header" if content.include?('Authorization') && content.include?('Bearer')
    
    if checks.length >= 2
      { ok: true, message: "✓ Structure valid (#{checks.join(', ')})" }
    else
      { ok: false, message: "✗ Missing key components" }
    end
  rescue => e
    { ok: false, message: "✗ Error checking structure: #{e.message}" }
  end
end

def main
  # Get parent directory (server/ruby)
  test_dir = File.dirname(__FILE__)
  parent_dir = File.dirname(test_dir)
  
  # Find all .rb files in parent directory
  ruby_files = Dir.glob(File.join(parent_dir, "*.rb"))
    .reject { |f| File.basename(f).start_with?('test_') }
    .sort
  
  if ruby_files.empty?
    puts "#{RED}No Ruby files found to test#{RESET}"
    exit 1
  end
  
  puts "#{BLUE}Testing #{ruby_files.length} Ruby sample files...#{RESET}\n"
  
  results = []
  ruby_files.each do |file_path|
    file_name = File.basename(file_path)
    puts "#{BLUE}Testing: #{file_name}#{RESET}"
    
    syntax = check_syntax(file_path)
    imports = check_imports(file_path)
    structure = check_structure(file_path)
    
    if syntax[:ok]
      puts "  #{GREEN}#{syntax[:message]}#{RESET}"
    else
      puts "  #{RED}#{syntax[:message]}#{RESET}"
    end
    
    if imports[:ok]
      puts "  #{GREEN}#{imports[:message]}#{RESET}"
    else
      puts "  #{YELLOW}#{imports[:message]}#{RESET}"
    end
    
    if structure[:ok]
      puts "  #{GREEN}#{structure[:message]}#{RESET}"
    else
      puts "  #{YELLOW}#{structure[:message]}#{RESET}"
    end
    
    all_ok = syntax[:ok] && imports[:ok] && structure[:ok]
    results << { name: file_name, ok: all_ok }
    puts ""
  end
  
  # Summary
  puts "#{BLUE}#{'=' * 60}#{RESET}"
  puts "#{BLUE}Summary:#{RESET}\n"
  
  passed = results.count { |r| r[:ok] }
  total = results.length
  
  results.each do |result|
    status = result[:ok] ? "#{GREEN}✓ PASS#{RESET}" : "#{YELLOW}⚠ PARTIAL#{RESET}"
    puts "  #{status} - #{result[:name]}"
  end
  
  puts "\n#{BLUE}Results: #{passed}/#{total} files passed all checks#{RESET}"
  
  if passed == total
    puts "#{GREEN}All files are valid! ✓#{RESET}"
    exit 0
  else
    puts "#{YELLOW}Some files need attention (likely missing dependencies)#{RESET}"
    puts "#{YELLOW}Run: bundle install#{RESET}"
    exit 1
  end
end

main

File: server/php/tests/test_all.php
<?php
/**
 * Test script to validate all PHP code samples.
 * This script checks:
 * 1. PHP syntax validity
 * 2. Import dependencies
 * 3. Basic structure validation
 */

// Colors for output
define('GREEN', "\033[92m");
define('RED', "\033[91m");
define('YELLOW', "\033[93m");
define('BLUE', "\033[94m");
define('RESET', "\033[0m");

function checkSyntax($filePath) {
    // Use php -l to check syntax
    $output = [];
    $returnVar = 0;
    exec("php -l " . escapeshellarg($filePath) . " 2>&1", $output, $returnVar);
    
    if ($returnVar === 0) {
        return ['ok' => true, 'message' => '✓ Syntax valid'];
    } else {
        $error = implode("\n", $output);
        return ['ok' => false, 'message' => '✗ Syntax error: ' . $error];
    }
}

function checkImports($filePath) {
    $content = file_get_contents($filePath);
    
    // Check for required imports
    $hasDotenv = strpos($content, 'Dotenv') !== false || strpos($content, 'dotenv') !== false;
    $hasCurl = strpos($content, 'curl_') !== false || strpos($content, 'CURLOPT_') !== false;
    
    $missing = [];
    if (!$hasDotenv) {
        $missing[] = 'vlucas/phpdotenv';
    }
    if (!$hasCurl) {
        $missing[] = 'curl extension';
    }
    
    if (!empty($missing)) {
        return ['ok' => false, 'message' => '✗ Missing dependencies: ' . implode(', ', $missing)];
    }
    
    // Try to check if composer dependencies are available
    $vendorPath = dirname($filePath) . '/vendor/autoload.php';
    if (!file_exists($vendorPath)) {
        return ['ok' => false, 'message' => '✗ Dependencies not installed (run: composer install)'];
    }
    
    return ['ok' => true, 'message' => '✓ Imports valid'];
}

function checkStructure($filePath) {
    $content = file_get_contents($filePath);
    
    $checks = [];
    if (strpos($content, 'Dotenv') !== false || strpos($content, 'dotenv') !== false) {
        $checks[] = 'Environment loading';
    }
    if (strpos($content, 'COURIER_API_KEY') !== false) {
        $checks[] = 'API key reference';
    }
    if (strpos($content, 'curl_') !== false || strpos($content, 'CURLOPT_') !== false) {
        $checks[] = 'HTTP client';
    }
    if (strpos($content, 'Authorization') !== false && strpos($content, 'Bearer') !== false) {
        $checks[] = 'Authorization header';
    }
    if (strpos($content, 'json_encode') !== false || strpos($content, 'json_decode') !== false) {
        $checks[] = 'JSON handling';
    }
    
    if (count($checks) >= 2) {
        return ['ok' => true, 'message' => '✓ Structure valid (' . implode(', ', $checks) . ')'];
    } else {
        return ['ok' => false, 'message' => '✗ Missing key components'];
    }
}

function main() {
    // Get parent directory (server/php)
    $testDir = __DIR__;
    $parentDir = dirname($testDir);
    
    // Find all .php files in parent directory
    $phpFiles = glob($parentDir . '/*.php');
    $phpFiles = array_filter($phpFiles, function($file) {
        return !strpos(basename($file), 'test_');
    });
    sort($phpFiles);
    
    if (empty($phpFiles)) {
        echo RED . "No PHP files found to test" . RESET . "\n";
        exit(1);
    }
    
    echo BLUE . "Testing " . count($phpFiles) . " PHP sample files..." . RESET . "\n\n";
    
    $results = [];
    foreach ($phpFiles as $filePath) {
        $fileName = basename($filePath);
        echo BLUE . "Testing: {$fileName}" . RESET . "\n";
        
        $syntax = checkSyntax($filePath);
        $imports = checkImports($filePath);
        $structure = checkStructure($filePath);
        
        if ($syntax['ok']) {
            echo "  " . GREEN . $syntax['message'] . RESET . "\n";
        } else {
            echo "  " . RED . $syntax['message'] . RESET . "\n";
        }
        
        if ($imports['ok']) {
            echo "  " . GREEN . $imports['message'] . RESET . "\n";
        } else {
            echo "  " . YELLOW . $imports['message'] . RESET . "\n";
        }
        
        if ($structure['ok']) {
            echo "  " . GREEN . $structure['message'] . RESET . "\n";
        } else {
            echo "  " . YELLOW . $structure['message'] . RESET . "\n";
        }
        
        $allOk = $syntax['ok'] && $imports['ok'] && $structure['ok'];
        $results[] = ['name' => $fileName, 'ok' => $allOk];
        echo "\n";
    }
    
    // Summary
    echo BLUE . str_repeat('=', 60) . RESET . "\n";
    echo BLUE . "Summary:" . RESET . "\n\n";
    
    $passed = 0;
    foreach ($results as $result) {
        if ($result['ok']) {
            echo "  " . GREEN . "✓ PASS" . RESET . " - {$result['name']}\n";
            $passed++;
        } else {
            echo "  " . YELLOW . "⚠ PARTIAL" . RESET . " - {$result['name']}\n";
        }
    }
    
    $total = count($results);
    echo "\n" . BLUE . "Results: {$passed}/{$total} files passed all checks" . RESET . "\n";
    
    if ($passed === $total) {
        echo GREEN . "All files are valid! ✓" . RESET . "\n";
        exit(0);
    } else {
        echo YELLOW . "Some files need attention (likely missing dependencies)" . RESET . "\n";
        echo YELLOW . "Run: composer install" . RESET . "\n";
        exit(1);
    }
}

main();

